File: src/leetcodejava/math/MajorityElement169.java
package leetcodejava.math;

import org.junit.Assert;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

/**
 * This is the solution of No. 169 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode.com/problems/majority-element/
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * Given an array of size n, find the majority element. The majority element is the element that
 * appears more than ⌊ n/2 ⌋ times.
 * <p>
 * You may assume that the array is non-empty and the majority element always exist in the array.
 * <p>
 * Example 1:
 * Input: [3,2,3]
 * Output: 3
 * <p>
 * Example 2:
 * Input: [2,2,1,1,1,2,2]
 * Output: 2
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class MajorityElement169 {
    @Test
    public void majorityElementTest() {
        int[] nums = {3, 2, 2, 2, 3};
        int result = majorityElement3(nums);
        System.out.println(result);
        Assert.assertEquals(result, 2);
    }

    /**
     * 计算每个字符出现次数
     *
     * @param nums 数组
     * @return map
     */
    private int majorityElement(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        int majorSize = nums.length / 2;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() > majorSize) {
                return entry.getKey();
            }
        }
        return -1;
    }

    /**
     * 找出一个数的个数大于数组中一半的长度
     *
     * @param nums 数组
     * @return int类型
     */
    private int majorityElement2(int[] nums) {
        int major = 0, count = 0;
        for (int num : nums) {
            if (count == 0) {
                major = num;
                count++;
            } else if (major == num) {
                count++;
            } else {
                count--;
            }
        }
        return major;
    }

    /**
     * 找出一个数的个数大于数组中一半的长度
     *
     * @param nums 数组
     * @return int类型
     */
    private int majorityElement3(int[] nums) {
        if (nums == null || nums.length < 1) {
            return -1;
        }
        return getMajority(nums, 0, nums.length - 1);
    }

    /**
     * 获取最大多元素
     *
     * @param nums  元素数组
     * @param left  左边位置
     * @param right 右边位置
     * @return 最多元素
     */
    private int getMajority(int[] nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }
        int mid = left + (right - left) / 2;
        int leftMajority = getMajority(nums, left, mid);
        int rightMajority = getMajority(nums, mid + 1, right);
        if (leftMajority == rightMajority) {
            return leftMajority;
        }
        int leftCount = 0;
        int rightCount = 0;
        for (int i = left; i <= right; i++) {
            if (nums[i] == leftMajority) {
                leftCount++;
            }
            if (nums[i] == rightMajority) {
                rightCount++;
            }
        }
        return leftCount > rightCount ? leftMajority : rightMajority;
    }
}

File: src/leetcodepython/array/majority_element_169.py
# encoding='utf-8'
'''
/**
 * This is the solution of No. 169 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode.com/problems/majority-element/
 *
 * The description of problem is as follow:
 * ==========================================================================================================
 * Given an array of size n, find the majority element. The majority element is the element that
 * appears more than ⌊ n/2 ⌋ times.
 *
 * You may assume that the array is non-empty and the majority element always exist in the array.
 *
 * Example 1:
 * Input: [3,2,3]
 * Output: 3
 *
 * Example 2:
 * Input: [2,2,1,1,1,2,2]
 * Output: 2
 * ==========================================================================================================
 * @author  zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


class Solution:
    def majority_element(self, nums: List[int]) -> int:
        '''
            大多数元素
        Args:
            nums: 数组
        Returns:
            元素
        '''
        num_dic = {}
        for num in nums:
            if num in num_dic:
                num_dic[num] += 1
            else:
                num_dic[num] = 1
        for key in num_dic.keys():
            if num_dic[key] >= len(nums) / 2:
                return key
        return -1

    def majority_element2(self, nums: List[int]) -> int:
        '''
            大多数元素
        Args:
            nums: 数组
        Returns:
            元素
        '''
        majority, count = 0, 0
        for num in nums:
            if count == 0:
                majority = num
                count += 1
            elif majority == num:
                count += 1
            else:
                count -= 1
        return majority

    def majority_element3(self, nums: List[int]) -> int:
        '''
            大多数元素
        Args:
            nums: 数组
        Returns:
            元素
        '''
        if not nums or len(nums) < 1:
            return
        return self.get_majority(nums, 0, len(nums) - 1)

    def get_majority(self, nums: List[int], left: int, right: int) -> int:
        '''
            获取大多数元素
        Args:
            nums: 数组
            left: 位置left
            right: 位置right
        Returns:
            int
        '''
        if left == right:
            return nums[left]
        mid = left + (right - left) // 2
        left_majority = self.get_majority(nums, left, mid)
        right_majority = self.get_majority(nums, mid + 1, right)
        if left_majority == right_majority:
            return left_majority
        left_count, right_count = 0, 0
        for i in range(left, right + 1):
            if nums[i] == left_majority:
                left_count += 1
            if nums[i] == right_majority:
                right_count += 1
        return left_majority if left_count > right_count else right_majority


if __name__ == '__main__':
    nums = [1, 2, 2, 2, 3]
    solution = Solution()
    num = solution.majority_element3(nums)
    print(num)
    assert num == 2
